### ISSUE TITLE
Keep multi-line formatting for destructuring assignment 

### ISSUE THREAD

**Issue by Brian-Gaffney** [prettier/prettier#2550]
I think multi-line destructuring should not be collapsed by Prettier.

Same concept as the [multi-line objects special case](https://github.com/prettier/prettier/blob/master/Rationale.md#multi-line-objects).

When doing destructuring (particularly nested) new lines can greatly enhance readability.

Example of current behaviour:
_Input_
```javascript
// Object with multiple lines uncollapsed by Prettier
const user = {
  name: {
    first: "Zoid",
    last: "Berg",
  },
};

// Similar destructuring assignment is collapsed by Prettier
const {
  name: {
    first,
    last,
  },
} = user;
```

_Prettier output_
```javascript
// Object with multiple lines uncollapsed by Prettier
const user = {
  name: {
    first: "Zoid",
    last: "Berg"
  }
};

// Similar destructuring assignment is collapsed by Prettier
const { name: { first, last } } = user;
```

**Comment by azz** [comment]
The output seems perfectly readable to me.

The special-casing of objects is more of a failing in prettier than a feature, I don't think it makes sense to spread that to more cases.

**Comment by Brian-Gaffney** [comment]
I agree that the above example is quite readable, I was only going for a minimal example case.

For larger and deeper destructuring assignments I find it harder to read.

**Larger example**
Input:
```javascript
const {
  name: {
    first,
    last,
  },
  organisation: {
    address: {
      street: orgStreetAddress,
      postcode: orgPostcode,
    },
  },
} = user;
```

Prettier output:
```javascript
const {
  name: { first, last },
  organisation: { address: { street: orgStreetAddress, postcode: orgPostcode } }
} = user;
```

If there's no interest in this change then I'm fine to drop it.
I found it a bit of a barrier to adoption at work where we have some very big props destructuring in React components and I thought others might have the same issue.

**Comment by azz** [comment]
Agreed, the input looks better.

Maybe a better rule would be "always break nested object patterns onto multiple lines", rather than looking at the source. Though that rule might have to only apply to variable declarations, else this will be an issue:


```js
function f({ data: { name } }) {}
//Out:
function f({
  data: {
    name
  }
}) {}
```

Which would likely be seen as a regression.

**Comment by Brian-Gaffney** [comment]
For React pure render functions I do find myself writing them similarly to your example output above.
Each destructed variable is on a new line.

```javascript
export default UserComponent ({
  name: {
    first,
    last,
  },
  organisation: {
    address: {
      street: orgStreetAddress,
      postcode: orgPostcode,
    },
  },
}) {
  return (
    <div>
      ...
    </div>
  )
};
```

I'd be fine with destructuring always using multiple lines but I imagine many others won't share that opinion.

Are you against referencing the source formatting because it's difficult or because it's a special case and against the general spirit of Prettier?

**Comment by azz** [comment]
> Are you against referencing the source formatting because it's difficult or because it's a special case and against the general spirit of Prettier?

The latter. But moreso because it makes prettier less predictable. Having to learn "tricks" to get prettier to print optimal code sucks. An example I run into all the time:

```js
Object.assign({
  foo: bar
}, baz);
```

Will be printed as

```js
Object.assign(
  {
    foo: bar
  },
  baz
);
```

When I usually want

```js
Object.assign(
  { foo: bar },
  baz
);
```

Which you can convince prettier to do by removing the line breaks, but not everyone knows that.

**Comment by alexrqs** [comment]
@Brian-Gaffney makes an excelent point, I use that destructuring a lot on react pure components and I'm a little tired of using `// prettier-ignore` plus I found myself on this scenario:

when you add or remove a key the diff will look like 
```diff
import {
  Foo,
  Bar,
-  Baz, 
}
```

but with prettier the hole line has changes 
```diff
- import { Foo, Bar, Baz } from './Qux'
+ import { Foo, Bar } from './Qux'
```

PS - Going for the minimal example here but keep in consideration real use cases like
``` {
  name: {
    first,
    last,
  },
  organisation: {
    address: {
      street: orgStreetAddress,
      postcode: orgPostcode,
    },
  },
}
```

**Comment by SimenB** [comment]
I just opened up #2685 as a dupe of this.

I'll reiterate my stance from that here.

>I would expect them to be treated the same - either both collapse, or both stay expanded.
>(Personally I would prefer both to be expanded, combined with #2068 being a thing you'd invoke sometimes)

**Comment by lewisl9029** [comment]
Maybe we can limit multi-line formatting to only destructuring statements with 2 or more props (possibly also make it configurable)? Similar to the ESLint `minProperties` option on their multi-line object rules (https://eslint.org/docs/rules/object-curly-newline#minproperties). This would still ignore existing formatting so prettier can format things in a consistent way.

**Comment by Panya** [comment]
@alexrqs yep, diff-friendly code style is a good thing.

**Comment by morgs32** [comment]
Not to sound overeager, but these sound like good reasons to me. Next I'd like to add a trailing comma. 

What do we need to do to keep the discussion moving forward. Perhaps it can be weaved into #74 ? Is it right to suggest that this format would be consistent with the way we (want to) treat object literals and lists?

**Comment by nocke** [comment]
If you also have less/css files in mind, consider scenarios like these:

```
body,div,dl,dt,dd,ul,ol,li,h1,h2,h3,h4,h5,h6,pre,code,form,fieldset,legend,input,textarea,p,blockquote,th,td,hr,button,article,aside,details,figcaption,figure,footer,header,hgroup,menu,nav,section {
  margin: 0;
  padding: 0;
}
```
Nobody want every of those selectors on a separate line. Which currently happens. (and using `// prettier-ignore` in too many places is no fun, either). Is there a way to generally suppress this for `.less`-files?

A linefeed right before `printWidth` gets reached (before more comma-separated selectors on the next line) would be good...

**Comment by lydell** [comment]
@nocke Your example is not very good for suggesting different formatting. You only do that reset _once,_ right? So it doesn't matter that much if that one rule gets a 20-line selector (in my opinion).

**Comment by sophistifunk** [comment]
Is the point of prettier to allow teams to agree on and enforce a style mechanically? Or is it to export the authors' particular preferences to the world at large?

If the former, then I think it would benefit one of:

* Respecting the original intent as the object syntax formatting does
* Using a max-item-count for both objects and destructure statements
* Allowing a character-count limit for single-line objects and destructure statement separate from line width

This is a big sticking point with some members of my team - I'm personally willing to pay the price of a few ugly destructures in order to get mechanical formatting, but I don't write the cheques.

**Comment by lydell** [comment]
>  is it to export the authors' particular preferences to the world at large?

Definitely not that one. Sometimes issues need a _lot_ of discussions before the best solution to go forward is found. Like the JSX ternary change. And sometime everyone ends up working on bugs or other features instead.

**Comment by suchipi** [comment]
> Maybe a better rule would be "always break nested object patterns onto multiple lines", rather than looking at the source. Though that rule might have to only apply to variable declarations - @azz 

I'm in favor of this- that said, I never use nested destructuring, partially because I prefer to keep prop shapes pretty shallow, and partially because I find it difficult to grok.

**Comment by nocke** [comment]
Back to two real-world javascript examples:
There are times, where you just want things as a big blob (my first example) –not wasting vertical space to scroll over– that's hardly ever to change:

```
const npmCommands = [
	access', 'add-user', 'adduser', 'apihelp', 'author', 'bin', 'bugs', 'c,
    cache', 'completion', 'config', 'ddp', 'dedupe', 'deprecate', 'dist-tag,
	dist-tags', 'docs', 'edit', 'explore', 'faq', 'find', 'find-dupes', 'get,
	...
    unlink', 'unpublish', 'unstar', 'up', 'update', 'upgrade', 'v', 'verison,
	version', 'view', 'whoami ];

```

...and then there are times, where you want everything on it's own line for sake of maintainability, even if dozens of lines:

```
const slugify: HashMap = {
	'£': 'pound',
	'¥': 'yen',
	'ß': 'ss',
	'à': 'a',
	'á': 'a',
	'â': 'a',
	'ã': 'a',
	'ä': 'ae',
	...
```
**Meaning:**
An `auto`-mode that detects (rather simple: are there „mostly“ linebreaks between the object/array members/selectors or not?) and then reinforces that on any minor inconsistencies... (as prettier should do).

**Comment by damassi** [comment]
Related https://github.com/prettier/prettier/issues/3101

**Comment by smallnamespace** [comment]
As a side note, ESLint has options to allow breaking these nested objects into multiple lines vs. keeping them consolidated.

At the minimum, it seems that prettier should be configurable to accept consolidated or broken-apart nested objects since the best style seem tricky to define except on a case-by-case basis.

**Comment by nocke** [comment]
> At the minimum, it seems that prettier should be configurable to accept consolidated or broken-apart
> nested objects since the best style seem tricky to define except on a case-by-case basis.

I would see (an option for) auto-detection as a good case-by-case basis. ([see above](https://github.com/prettier/prettier/issues/2550#issuecomment-336048119)) Do not rely on length or any other assumption, rely an what's there (and then consolidate, if the apparently chosen style is only „mostly“ there...).

By the way, js-beautify [has something like that](https://github.com/beautify-web/js-beautify#options), appending a ` ,preserve-inline`

    "brace_style": "collapse,preserve-inline",

tested, and it works pretty nicely...

**Comment by damassi** [comment]
@nocke, yup that would be a dream

**Comment by machineghost** [comment]
Regarding:
>Meaning:
>An auto-mode that detects (no rocket since, rather simple: are there „mostly“ linebreaks between the object/array members/css selectors or not?) and then reinforces that on any minor inconsistencies... (as prettier should do).

and:
>I would see (an option for) auto-detection as a good case-by-case basis. (see above) Do not rely on length or any other assumption, rely an what's there (and then consolidate, if the apparently chosen style is only „mostly“ there...).

I'm not in any way affiliated with the project, but based on my understanding of it that would be impossible.  Specifically I'm referring to the "detects ... are there mostly linebreaks between ..." part and the "rely on what's there" part.  As I understand it prettier reduces your code to an AST first; that's part of the "magic" of how it can format files consistently.  And doing that amounts to discarding all formatting that was there previously.

In other words, Prettier can't do anything based on the way you had formatted something, because it is designed to throw out everything you had before and just generate a consistent result.

I think the real fix here would be to give users more control over how Prettier output things _every time_.

**Comment by j-f1** [comment]
> In other words, Prettier can't do anything based on the way you had formatted something, because it is designed to throw out everything you had before and just generate a consistent result.

That’s not entirely true — Prettier does have ways of accessing the original source, and that’s used to keep multline objects multiline and preserve (single) blank lines. There is a feature request somewhere for an option to disable this and only use the AST, though.

**Comment by josephfrazier** [comment]
> There is a feature request somewhere for an option to disable this and only use the AST, though.

Here it is, for reference: https://github.com/prettier/prettier/issues/2068

**Comment by narkowicz** [comment]
As a new user testing Prettier in a fresh project, the newline behaviour around all destructuring (including import statements) and arrays, along with the closely related issues regarding jsx #3101 and json #2716, are the only deal-breakers preventing adoption.

Our own reasons for using multi-line formatting in imports, destructures, objects, arrays, etc. relate mainly to supporting simple / automatic re-ordering and easy addition and removal of props, all of which help maintain clean git history (i.e. when used with `trailingComma`).

As mentioned in https://github.com/prettier/prettier/issues/3101#issuecomment-357739674 code reviews benefit from multi-line props - this extends to merge operations - and the same applies when destructuring.

Without this configuration option, it's possible we would see _more_, not less, formatting related changes in our git history which is at odds with our own motivations for implementing Prettier.

FWIW the only time I **_wouldn't_** prefer a multi-line destructure is in an inline arrow function such as:

```js
[
  {
    key: 'value',
  },
].map(({ key }) => key);
```

...but given the choice I would rather have the option to _always_ enforce newlines when destructuring than support edge cases like this.

The running suggestion in https://github.com/prettier/prettier/issues/3101#issuecomment-355279030 of defining the max number of props per line seems like it could carry across for all destructuring and import statements as well. i.e. a value of `0` would always enforce multi-line, `1` allows for a single inline destructure (as in example above), and `3` inline props seems to be the magic number for those not attached to blanket enforcement. Could this be combined with another option to respect `intent` as per https://github.com/prettier/prettier/issues/2550#issuecomment-348442621 ?

---

As an aside, the current destructure formatting seems inconsistent with the formatting applied when constructing an object- which appears to respect the author's intent. i.e. the following passes through formatting as-is:

```js
const a = 0;
const b = 1;
const c = { a, b };
const d = {
  a,
  b,
};
```

(although this is not the case with arrays...)

**Comment by azz** [comment]
Could we just _always_ break onto multiple lines if there's nested destructuring, and keep it as it is for flat destructuring?

**Comment by narkowicz** [comment]
If it warrants a configuration option I'd love to see a setting that supports always breaking.

@azz to illustrate my preference, compare the following diffs:

```diff
# Breaking nested destructure only
-const { a, b, c } = someObject;
+const {
+  a,
+  b,
+  c,
+  d: {
+    e,
+  },
+} = someObject;
```

```diff
# Always breaking (clear addition, minimal delta)
const {
   a,
   b,
   c,
+  d: {
+    e,
+  },
 } = someObject;
```

Meanwhile changes in another branch now need to be merged with the nested destructure formatting - each case results in a conflict - compare the two diffs:

```diff
# Breaking nested destructure only (conflict must be resolved manually)
-const {
-  a,
-  b,
-  c,
-  d: {
-    e,
-  },
-} = someObject;
+const { a, b, c, f, g } = someObject;
```

```diff
# Always breaking (easily resolved with merge tool / editor by accepting both changes)
const {
   a,
   b,
   c,
-  d: {
-    e,
-  },
+  f,
+  g,
 } = someObject;
```

**Comment by azz** [comment]
We can't _always_ optimize for diffs. Imagine if we printed _every_ function like this:

```js
function foo(
  a,
  b
) {}
```

We don't want to do that, and your example is similar. e.g.:

```js
const {
  a,
} = b;

// or

const f = ({
  a,
}) => a;
```

**Comment by shtanton** [comment]
I like having pretty much everything as expanded as is possible, my method to get prettier to do this was to set printWidth really low.  Mine is at 20.  This means it expands nearly everything as much as possible

**Comment by Slapbox** [comment]
I'd really like to use Prettier _AND_ have readable code, which as I understand it is the goal of this project.

The current destructuring output is just atrocious. Not only is it unreadable, but it makes debugging difficult. I can't simply comment out a property now. I have to delete it, or manually expand the code, and hope I remember to replace it later.

Can we please get resolution on this? It's a serious problem with an otherwise excellent tool.

**Comment by azz** [comment]
> Since this has been open for nearly 8 months without comment from the owner, I am guessing this issue will not be resolved.

This is not true, while James is is author of this project, there are many maintainers (collaborators) capable of resolving the problem. 

What we are lacking is a heuristic to use to satisfy this problem without introducing style regressions. To help with this, would you be able to share snippets of code that look bad when Prettier formats them?

**Comment by Slapbox** [comment]
@azz you're right about the specific point on the comment from owner. I should not have made this point. That being said,the issue has been open for 8 months and the idea, while being discussed, it, doesn't seem to have much support from the maintainers of this library.

It's not simply a matter of code that looks bad. It's that this makes the code hard to work with. That being said, any import of a sufficient number of properties is going to look bad.

I've removed Prettier from my configuration. Is there a place I can easily run code through prettier to demonstrate? In the meantime, here's a rough approximation (probably wrong):

```
    const {
      property1,
      property2,
      property3,
      property4,
      property5
      property10
    } = this.props;
```
>> Becomes >>
```
    const { property1, property2, property3, property4, property5} = this.props;
```

The readability here is lacking only slightly. The much bigger problem with this is that the code is not pliable in this form. If I want to comment out an item, this is now kind of a chore. During a refactor of older/dirtier code, you may often want to comment out these items instead of deleting them for easy reversibility.

```
    const {
      property1,
      property2,
      property3,
      // property4,
      property5
    } = this.props;
```

**Comment by suchipi** [comment]
@Slapbox you can use the online prettier playground here: https://prettier.io/playground/

The "Copy markdown" button in the bottom right is very useful when sharing results on GitHub.

**Comment by Slapbox** [comment]
@suchipi thanks I actually did use that, but I forgot to update my text.

**Comment by dlombardi** [comment]
Can a maintainer please signify some forward momentum on this?  Multi-line destructuring deserves a thought out response.  It's the only reason we do not use prettier on my team.  Desctructuring nested objects is indeed not pliable and is frankly a chore to read.

**Comment by lydell** [comment]
> It's the only reason we do not use prettier on my team.

Everyone says that in every issue :)

**Comment by damassi** [comment]
@dlombardi - Sometimes its good to just try things out for a while and suppress judgement, and acknowledge that it can take getting used to. All of the preferences that come with having written hand-crafted code over the course of a career are tough to let go of, but if you can manage it its shocking how much friction Prettier removes, especially with VSCode `formatOnSave` turned on. (How much time did I used to spend moving copy / paste code around? Now `cmd+s` does all that for me, instantly. Its amazing.) 

I was way, way on the fence due to destructuring and multi-line JSX props and a few other things but once I gave it an honest spin over a substantial codebase I can't believe I ever lived without this tool. You'd be doing your team a great favor by implementing it rather than blocking, or at least taking it for a longer test-run. I can almost guarantee you that your issue with being able to comment out destructured props will very quickly fade as you get into the flow of things and start understanding just now much it assists (and speeds up) the code-writing process.

**Comment by machineghost** [comment]
@dmassi I think that was an excellent defense of all of Prettiers lack of customization, not just destructuring objects, and we both agree that Prettier is really great.

However, I disagree that with the idea that (if you'll forgive me for re-stating what you said in a crass way) "there's nothing wrong with Prettier, the problem is you for not liking how Prettier does things, so just try it and the problem (ie. you) will be solved."  And I also don't think something being good is an argument against making it better.

Prettier would be a better library for everyone if it offered more customization options in general, and one for multi-line object properties in particular.  And I say this as someone who *has* suppressed judgement, and *has* used the library despite it doing things I'm uncomfortable with.

Is my life better with Prettier, in spite of those uncomfortable things?  100%, I'm not going back. But that's not the same thing as saying Prettier is the best tool possible and adding more customization would make it worse.

As an outsider looking in at the lack of progress on tickets like this one it certainly _feels_ like the lack of customization comes from resistance from the maintainers, and I think that's sad because no one developer or software team knows what is right (or even what is right modulo Prettier's five-odd customization options) for every other developer/team.

**Comment by duailibe** [comment]
@machineghost 

I invite you to read our docs section about [our philosophy on options](https://prettier.io/docs/en/option-philosophy.html) which answers your post more in depth.

**Comment by machineghost** [comment]
@duailibe I've read it, and I feel my post speaks to its deficiencies.    Prettier is two things conflated:  an _incredible_ formatting a tool, and an opinion on how everyone should write their code.  My opinion is that it could be a more incredible tool by easing up on the idea that Prettier's maintainers know how everyone's code should be written.

If the desire is to codify a standard that prettier can easily (in a technical sense) offer one, while still allowing shops to customize differently form that standard if they desire.  Babel offers presets, that's great as they put every Babel-using developer on the same page.  But Babel certainly doesn't try and stop anyone from adopting a module that's not in that preset, and I don't think "The Javascript Code Formatting Tool" should be any different from "The Javascript Transpiling Tool".

**Comment by dlombardi** [comment]
Not to be obtuse but I've still not heard a rationale behind the current destructuring method and why it is not instead multi-line.

Event better,

> Could we just always break onto multiple lines if there's nested destructuring, and keep it as it is for flat destructuring?

Every JSON formatter prints structures in multi-line formats for what I assume is for its readability.  Same kind of formatting is done for GraphQL queries.  If destructuring is effectively getting (setting variables to) data from an object shouldn't we have the same consistent standard that these popular tools encourage?

**Comment by ericanderson** [comment]
Having now put in 2 PRs into this code base... this thing is complex. If there were more flags I cannot fathom how difficult it would be to maintain. It would certainly be hard for an outsider like me to make a contribution.

I would like to see better heuristics on this too, but now I am thinking about complexity, and what would those rules look like? What are all the edge cases?

I really like the original code in https://github.com/prettier/prettier/issues/2550#issuecomment-318354352

That said, if the property names matched the variables, I would probably want it like this:

```ts
const {
  name: { first, last },
  organisation: { 
    address: { street, postcode }
  }
} = user;
```

Anyway, the only way forward is to make concrete proposals. Complaining will accomplish nothing.

So, if we put rules to it (I'm going to call these objects , but I also mean destructured object definitions).

An object is "simple" if any are true:
* It has less than N property shortcut entries only
* It has less than M property/variable pairs only (lets make mixed shortcut and pairs "complex")

In the case of object construction, this would also mean that computed values imply the wrapping object is complex.

An array is "simple" if any are true:
* It has less than O values and those values are all variables
* It has 1 value (computed or not) and that value fits on a single line

Rules: 
* If an object literal is simple, single line
* Otherwise its multiline

This matches my above example.

It would convert:
```ts
const {
  name: {
    first,
    last,
  },
  organisation: {
    address: {
      street: orgStreetAddress,
      postcode: orgPostcode,
    },
  },
} = user;
```

into 
```ts
const {
  name: { first, last }
  },
  organisation: {
    address: {
      street: orgStreetAddress,
      postcode: orgPostcode,
    },
  },
} = user;
```
where N === 5, M === 2

I *think* this is doable and works for both destructuring and object literals.

If N, M, and O are configurable, then the people who want everything broken per line get their way.

Lets discuss proposals please.

**Comment by duailibe** [comment]
@machineghost 

To be perfectly honest, Prettier has no interest in how people write their code. Users can very well write everything in a single line and Prettier won't say a single word about that, it'll simply format it :)

More seriously though, it's a misconception that the maintainers want to dictate style of everyone's code.. there are several examples of: things Prettier prints that isn't aligned on how we personally would write our code; options Prettier now accepts that almost none of us use and yet we coded an maintain them; and more examples of things we happily do, maintain or fix that don't use at all.. all this to say we have 0 interest in pushing "our" style..

> Babel offers presets, that's great as they put every Babel-using developer on the same page. But Babel certainly doesn't try and stop anyone from adopting a module that's not in that preset, and I don't think "The Javascript Code Formatting Tool" should be any different from "The Javascript Transpiling Tool".

I can't understand this, Babel and Prettier are tools with completely different goals, I can't see how they are similar at all.

@dlombardi 

From https://github.com/prettier/prettier/issues/2550#issuecomment-368275296:

> What we are lacking is a heuristic to use to satisfy this problem without introducing style regressions. 

We never said the current output is better but we just can't change it to simply always print multiline. We'd have this whole argument again for people that want to inline destructuring in some cases.

**Comment by dlombardi** [comment]
@ericanderson's heuristic groundwork has my full support

**Comment by dlombardi** [comment]
@ericanderson heuristic groundwork has my full support 👍

**Comment by ericanderson** [comment]
I want to think through a few more cases.

N=4, M=2, O=4

```js
let foo;

// #1
foo = { bar };

// #2
foo = { bar, baz, bloop, blam };

// #3
foo = { 
  bar, 
  baz, 
  bloop, 
  blam,
  bong,
};

// #4
foo = {
  bar: { baz, bloop, blam, bong}
}

// #5
foo = [ baz, bloop, blam, bong ];

// #6 
foo = [
  bar, 
  baz, 
  bloop, 
  blam,
  bong,
];

// #7
foo = {
  foo: doSomething()
}

// #8
foo = { foo: doSomething } // reference, not call

// #9
foo = [ isBaz ? getBar() : getBloop() ];

// #10
foo = [ 
  {
    bar: [ isBaz ? getBar() : getBloop() ],
  }
];

// #11
foo = [ 
  {
    bar: [ baz, bloop, bam ],
  }
];  
```

**Comment by ericanderson** [comment]
```js
// #12
foo = [
  {
    bar: [ 
      isBaz ? getBar() : getBloop(),
      5,
    ],
  }
];

**Comment by ericanderson** [comment]
@duailibe @j-f1 @azz @suchipi thoughts on the proposal? I might take a stab at it if there is support

**Comment by suchipi** [comment]
It's pretty complicated, but it seems like a good heuristic. However, I think that:
* N, M, and O should *NOT* be configurable
* This should only apply to object destructuring, not object literals.

**Comment by suchipi** [comment]
Actually, thinking about it some more, I think N, M, and O should be based more on print width and less on arbitrary numbers, otherwise we will introduce confusing style regressions...

**Comment by suchipi** [comment]
How about:

An object is "simple" if it only contains shorthand properties, and is "complex" otherwise.
An array is "simple" if it contains no computed properties.

Rules:
  If an object is simple, it's single-line if that fits under the print width
  Otherwise, it's multi-line

**Comment by ericanderson** [comment]
Interesting. For completeness I will paste out all my examples again but for your proposed rules @suchipi 

Im going to tweak it to say that an array is "simple" if it contains no computed properties OR only 1 computed property

```js
let foo;

// #1
foo = { bar };

// #2
foo = { bar, baz, bloop, blam };

// #3
foo = { bar, baz, bloop, blam, bong };

// #4
foo = {
  bar: { baz, bloop, blam, bong}
}

// #5
foo = [ baz, bloop, blam, bong ];

// #6 
foo = [ bar, baz, bloop, blam, bong ];

// #7
foo = {
  foo: doSomething()
}

// #8
foo = {
  foo: doSomething
} // reference, not call

// #9
foo = [ isBaz ? getBar() : getBloop() ];

// #10
foo = [ 
  {
    bar: [ isBaz ? getBar() : getBloop() ],
  }
];

// #11
foo = [ 
  {
    bar: [ baz, bloop, bam ],
  }
];  

// #12
foo = [
  {
    bar: [ 
      isBaz ? getBar() : getBloop(),
      5,
    ],
  }
];
```

**Comment by ericanderson** [comment]
I tweaked your rules because I wouldnt want 9 to render as:
```js
foo = [
  isBaz ? getBar() : getBloop()
];
```

Although I guess I don't mind that either. So your original rules seem reasonable.

**Comment by ericanderson** [comment]
Taking the OP's larger example we get

```js
const {
  name: { first, last },
  organisation: {
    address: {
      street: orgStreetAddress,
      postcode: orgPostcode,
    },
  },
} = user;
```

**Comment by suchipi** [comment]
I'm a bit confused; all your examples are object and array literals, not destructuring patterns. I thought this issue was about destructuring patterns

**Comment by ericanderson** [comment]
It’s both. The complaint includes that they behave differently.

**Comment by duailibe** [comment]
Yeah people complain that they behave differently (as in object literals you can force multiline whereas when destructuring you can't) but the solution is not to make them behave the same, but figure out heuristics of when to break the destructuring.

I agree with @suchipi that we should not change object (or array) literals at all. Only destructuring and I'd start with object destructuring first..

**Comment by ericanderson** [comment]
That’s fair. I think @suchipi s rules work nicely for the object restructuring case though. Still happy to address a fix that is much much smaller in scope

**Comment by mssngr** [comment]
If this issue gets resolved and PR merged in for treating destructured objects the same as object literals, I will get on the Prettier bandwagon again. As it is, now, it makes my code look like mush. :/

**Comment by nescalante** [comment]
When trying to implement prettier in our project we fell into this issue and we think this is a big loss for us.

It was mentioned before but there is a big loss in:
- readability for objects in [most of the cases](https://github.com/prettier/prettier/issues/2550#issuecomment-318354352)
- worst change tracking in GitHub since when you want to add a prop, you have to modify [the entire line](https://github.com/prettier/prettier/issues/2550#issuecomment-359409325) with props that you are not modifying instead of just adding what you need

@azz, is this proposal something that is still being considered?

**Comment by duailibe** [comment]
@nescalante If the issue is open, it's still being considered

**Comment by narkowicz** [comment]
Has there been progress on this?

I just entered @Brian-Gaffney's original example in the playground (with default settings) and got the desired result (apologies for the lengthy post):

**Prettier 1.11.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6VACA8gIwFZxgwYDuAljABYYC2ArgDYxkAODcGDZUcAzhnSiQGDAIYtecACYYcATwwAFAE5wYzOMoA6Q6L2J1JyjAF4MwHRgxRRNOEnOWrGAGZll+h1pAAtCGSlvABonKzFPDG8AIU0Ac2CnAF8QqESAbh0ddAwAZTIaMjFjKT4YZToiOmVuWIxRXl4yWKg7WAwyfmExCWlZBRU1DW1dKH1HKCsbOwcLCec3DxgU5056paSUxNMBIzSQIJAIFmY9ZFAWatgAdQCqZAAOAAYDmFEcG6k7pAAmA8aoWLsACKdAg8GQLlEDEkL2UokKNQAwhAaDRRMgQFBoHB9iAcHCwABrNQ5FiiMA1ZBlOhwA54XgADyiBOJMBytjgABluDikJDobSQGSPJoMTg3nIGNBcZIClTyoLDHAACpvXgQqEwkDcIwwFSiWJojUCg6qACOdHccH1hvRfM1gouEEkVzhLAxFz4mgAbjiDqJlMoICRFAGEOqUKJvf4pLi4QC4DlXsoYMhniB47E4ABRKCxpDpqQQMDGrVrUuCiB0GAsavfCuJRJAA)
```sh
--range-end 0
--range-start 0
```

**Input:**
```jsx
// Object with multiple lines uncollapsed by Prettier
const user = {
  name: {
    first: "Zoid",
    last: "Berg",
  },
};

// Similar destructuring assignment is collapsed by Prettier
const {
  name: {
    first,
    last,
  },
} = user;
```

**Output:**
```jsx
// Object with multiple lines uncollapsed by Prettier
const user = {
  name: {
    first: "Zoid",
    last: "Berg",
  },
};

// Similar destructuring assignment is collapsed by Prettier
const {
  name: {
    first,
    last,
  },
} = user;
```

Followed by his extended example which appears unchanged:

**Prettier 1.11.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEwA6VOZQCGAtnEtngQQGYCWAThgDRXUA2RLbAvq-pggMA5kSh00RGHWgVcAgkQAmShnDRo5bapgxq4MCkOEBlGPpgBBFWo38dBAA4QMkJeUEiACi5hu49tR8vPw8mAC8mACuaHAMANwgzCAQjtLoyKBEDAwQAO5e2QhoyCBEAG4QdEpJZRjINETssckARgxEYADWBiaOnXRQwsjmUQEgShBgDU0tIIOxDDBeHcIkRDPN4wBWaAAeAEId3b2kcAAyg3CbcxBRMI73AEw34-1McaWtRK0AnuzQWqOBiDGAAdWqMAAFsgABwABmSwJccDBHUcpWB6ji5WuyQ6QzgAFEoDUkIiQAThHAzNkYMgKWoAI5RRhwFZENYbJCNLbJWIkOgjBhjfmDYTsOAARSiEHgr2SMB+EKU0OQT0VHTo7HFAGEICR1qUoNA8SAYnAACo-Eo82ZwHg8IA)
```sh
--range-end 0
--range-start 0
```

**Input:**
```jsx
const {
  name: {
    first,
    last,
  },
  organisation: {
    address: {
      street: orgStreetAddress,
      postcode: orgPostcode,
    },
  },
} = user;
```

**Output:**
```jsx
const {
  name: { first, last },
  organisation: { address: { street: orgStreetAddress, postcode: orgPostcode } }
} = user;
```

But building on that example to include the object being destructured seems to produce the desired result 😕 :

**Prettier 1.11.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEBXNcBOmAvJsADpSaZQCGAtnEqRVVQGYCW+GTA5AFoQOAE14AaFqwA2NHpl4AhAgHNxkgL4TKmCPmU0oHNDRgdoTctqo1hw-HDRoLk1pgz24MPgEYAbABYABkwAJQ4ANwI0EThMABF8CLg1K1YABwgMSGFGeQAmAGY8-zyU100NLXUAbgoKSCgMZm1aBmdUzm4YLVcZDB7MCu1dfUNjU3Nm1xs7Bycp1zcYDy8dPQBlZbhPAEFbe0cB1wysiBymEYAFTJhsuCPBgaH1Yhw8fGqQMRAINInG5CgGj4fAQADul2BCDQyBANHCQmEXzhGGQbBoUjw3wARvgaGAANaedZpfEcKDKZDLbD3EDCCBgNEYrEgcnvGCXPHKOg0JmY2kAKzQAA8FHjCcT6HAADLkuB8lkQbAwNLKvIK2mk7gEWHYmjYgCeUmgyLSiVgAHURDAABbIAAcgW+ZsycAteLSsLNDgIkWReIpcAAolAkUgnSAA8o4JtgTBkBH7ABHbBcOCcmjc3lIdH8754OgcKn4Gn58nKKRwACK2Ag8A13xg+qtwltyDyjbxHCk5YAwhA6DzYVBoPLvrg4AAVfUwnPMuDqdRAA)
```sh
--range-end 0
--range-start 0
```

**Input:**
```jsx
const user = {
  name: {
    first: 'Zoid',
    last: 'Berg',
  },
  organisation: {
    address: {
      street: '1640 Riverside Drive',
      postcode: '23242',
    },
  },
};

const {
  name: {
    first,
    last,
  },
  organisation: {
    address: {
      street: orgStreetAddress,
      postcode: orgPostcode,
    },
  },
} = user;
```

**Output:**
```jsx
const user = {
  name: {
    first: "Zoid",
    last: "Berg"
  },
  organisation: {
    address: {
      street: "1640 Riverside Drive",
      postcode: "23242"
    }
  }
};

const {
  name: {
    first,
    last,
  },
  organisation: {
    address: {
      street: orgStreetAddress,
      postcode: orgPostcode,
    },
  },
} = user;
```

Is there a reason for the inconsistency between the two extended examples? Is there cause for celebration?

**Comment by ikatyang** [comment]
@narkowicz sorry for the confusing, it's caused by a bug in our playground (#4273), as a temporary workaround you'll have to set `--range-end` to something like `999` to include all the content.

**Comment by duailibe** [comment]
I have a PR (#4267) the addresses some of this. Can you try it out on the [PR's playground](https://deploy-preview-4267--prettier.netlify.com/playground/#N4Igxg9gdgLgprEAuc0DOMAExMFc1wBOS2mAZgJaEYByAhgLZwA0mANnbY3JgL5+YAvJhgALCmgB0AB0IRpaANwAdKKshQM2VZkxRuJHJWoxWHLb2Y7MEQgHM6UCXRgVohzHQAmXwnDRoHhh+cDAktnYAyjAhMACCPn4BrNIQGJBecOH2AAppMBk8-Lyq-ML4RCpqUGS4UGCu0OQAFDheLnQe+kwCvACU2CXVGloAqgSEAMIQDKlQCFjCtfWNUK3W3VmkxhhmnFiW1hGOzqse3r7+gaTBcKHZUTF38YlXKfmFD3npEJm9VlB+tooLo-DBcIRqrwqup0FgcHRWAAjVhgVheDxFATCNAzOAAeSRACs4A0qiBmCB5Ks0MhQHRCHIAO45BkIWkoOgANwgFC8FJA+2QZDobAIlKRhDoYAA1qFItJpRQoHZkDFcCwQF4IGBhaLxSBlRMYDkpXYGHQ9WLNUS0AAPABCUtl8u4ABllXArQaILgYNI-QAmb2axXUIjIEBIuhIgCebGgAtkypgAHU+WJkAAOAAMlNkaTgqal0kjsn8RC5XspfgAjrgqHBTXRzZakCLrZSCAwKGrCBqu8q7Gw4ABFXAQeAhykwGPpryZpCBmdSihsIfTBgWyNQaDVkAVAAqMY5HYIvF4QA) which addresses some of the cases reported here.

**Comment by Favna** [comment]
I just found this issue when looking for solutions to my current code formatting conflict between eslint and prettier and I'd like to know if there's anything to resolve it. 

What I want to be able to have is:

```js
const {deleteCommandMessages, roundNumber, stopTyping, startTyping } = require('../../util.js');
```

but what prettier does for me right now is:

```js
  {
    deleteCommandMessages,
    roundNumber,
    stopTyping,
    startTyping
  } = require('../../util.js');
```

Now this entire thread is kind of confusing to me because from the title I would presume it is already possible to NOT have multi-line formatting, but then again I see no option about this in the docs at all. So, can I tell Prettier to not put each destructured property on a new line?

[Here is my current eslintrc file](https://github.com/Favna/Ribbon/blob/master/.eslintrc.js)

**Comment by lydell** [comment]
@Favna We recommend disabling all ESLint rules that conflict with Prettier, for example using [eslint-config-prettier](https://github.com/prettier/eslint-config-prettier)

**Comment by Favna** [comment]
Sadness. So that basically means I cannot use Prettier because there is no option to disable that new line formatting.

**Comment by j-f1** [comment]
@Favna why can’t you disable the ESLint rule?

**Comment by mssngr** [comment]
@Favna, from what I understand, Prettier is an _opinionated_ formatter. So, you can use it, but it's definitely going to format your destructured object with that many properties to a new line, and I don't think they have plans to add additional configuration to change that.

**Comment by Favna** [comment]
@j-f1 It's not an eslint rule that causes the issue. Eslint doesn't care if you put destructured objects across 1 line. The closest thing to it is [object-property-newline](https://eslint.org/docs/rules/object-property-newline) but even that does not affect this newline behavior. 

I'd say @mssngr summed it up nicely then. Since I cannot conform myself to the opinionated formatting that Prettier uses it sadly is not for me.

**Comment by ljharb** [comment]
It’s unfortunate that prettier is actually two things conflated into one: a performant, AST-based code formatter; and an opinionated style guide. The former is unique and valuable, but the latter is just another in a cluttered category.

**Comment by jcollum** [comment]
OK this is still being discussed right? Hasn't been implemented? A team member asked for multiline destructuring and I use Prettier and this is the first google hit. So if it hasn't been implemented then I'll need to work around it.

**Comment by duailibe** [comment]
@jcollum We're now breaking when there's nested destructuring.. see the [blog post of 1.12](https://prettier.io/blog/2018/04/11/1.12.0.html#break-objectpattern-s-with-nested-objectpattern-s-4267-https-githubcom-prettier-prettier-pull-4267-by-duailibe-https-githubcom-duailibe).

We're not sure this does address all the cases discussed here, so the issue is still open.

**Comment by lydell** [comment]
> This issue page is gigantic!

They tend to be on popular projects.

**Comment by finalclass** [comment]
Weirdest thing ever:

Before:
```js
const { a } = {};
```
After:
```js
const {
    a
} = {};
```

Seriously fix that or at least add an option to disable formatting of destructuring.

**Comment by j-f1** [comment]
@finalclass I can’t reproduce that on the playground:

**Prettier 1.12.1**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAExMENMC+mAvNgQNwA6UIANCBAA4wCW6youATlxAO4AFbgjTIQuAG4QWAEzogARl1xgA1nBgBlRipZQA5shhcArnHp60cLjAHL9AW1zIAZrgA2V+gCs0ADwAhZTUNTVwHOAAZPThXDy8QHS4rLjEFXAUAT3doeUYuPRgAdVkYAAtkAA4ABnp8iCsi5UYxfLgUiVj6LjgARxMWHrtcR2ckN09zECsHFiNTKbQ9fXc4AEUTCHg4yfoYDJKZcuQAJj3lFndlgGEIBycxKGgukBMrABUM0XH4uAICIA)

**Input:**
```jsx
const { a } = {};

```

**Output:**
```jsx
const { a } = {};

```

Can you create a reproducible example of your issue there and post it as a new issue?

**Comment by finalclass** [comment]
So there must be something wrong with prettier vscode plugin:

![ezgif-1-d9de541b60](https://user-images.githubusercontent.com/1343930/39757379-f0f07bc6-52cc-11e8-8fef-b3e834ab8e72.gif)

**Comment by azz** [comment]
@finalclass could you raise this over at https://github.com/prettier/prettier-vscode/issues? While it is similar to this issue, the root cause is mostly likely very different.

**Comment by finalclass** [comment]
@azz Sure no problem!

**Comment by j-f1** [comment]
@finalclass I’d recommend opening an issue on the plugin’s repository.

**Comment by garrettberg** [comment]
Regardless of the specific rules, object destructuring should be treated exactly the same as object assignment. I think the examples below show how odd it is to treat them any differently:

Before:
```js
const example = {
  hello: "world",
  foo: "bar",
};

const {
  hello: world,
  foo: bar,
} = example;

const shorthandExample = {
  definedEarlier,
  anotherOne,
};

const {
  helloWorld,
  fooBar,
} = simpleDestructureExample;
```

After: 😢 
```js
const example = {
  hello: "world",
  foo: "bar"
};

const { hello: world, foo: bar } = example;

const shorthandExample = {
  definedEarlier,
  anotherOne
};

const { helloWorld, fooBar } = simpleDestructureExample;
```

Right now, object assignment maintain multiline if that's how the code was written. But if a new rule is introduced in the future (e.g., to keep shorthand object assignment on a single line) then the same should apply to the destructuring (in this example, simple destructuring without aliases or default values).

**Comment by waynebloss** [comment]
Does anyone know of a different formatter that has this option?

**Comment by j-f1** [comment]
Now that it’s less of a hassle to collapse expanded objects, perhaps this would be OK to do.

**Comment by j-f1** [comment]
cc @prettier/core on my above comment ⬆️

**Comment by vjeux** [comment]
I’d rather try to find cases where we can avoid respecting \n in objects rather than trying to add more of it.

**Comment by leojnxs** [comment]
> The output seems perfectly readable to me.
> 
> The special-casing of objects is more of a failing in prettier than a feature, I don't think it makes sense to spread that to more cases.

`> The output seems perfectly readable to me.`
- Because is sooooo readable use 6 lines to express what you just need 1 line.
For me, prettier should brake only code that exceeds line max length limit

**Comment by duailibe** [comment]
> For me, prettier should brake only code that exceeds line max length limit

@leojnxs Sure but that's not how Prettier works.

Please be more respectful

**Comment by leojnxs** [comment]
> @leojnxs Sure but that's not how Prettier works.
> 
> Please be more respectful

@duailibe Sorry for hurt your feelings. I'll be more respectful next time. Get better!

**Comment by leojnxs** [comment]
> @leojnxs Sure but that's not how Prettier works.
> 
> Please be more respectful

@duailibe I think everybody here knows how prettier works.
But we don't agree with everything and we are here trying to suggesting things to make it even more better.

Sorry if I was not respectful, `Sure but that's not how Prettier works.` is not the best example of respectful answer.

Sry for hurt your feelings, get better. Love Prettier. Love you. Have a nice day!

**Comment by waynebloss** [comment]
It seems like different people have different preferences for things. Crazy, I know... Maybe a highly opinionated formatter that only offers the bare minimum configuration options isn't the right tool for the vast majority of users.

It's certainly easier to not offer this option and the project maintainers can do what they want, but if enough people don't like it they'll jump ship as soon as something better comes along that offers more freedom.

Personally, I'd like more options for how my code wraps.

**Comment by j-f1** [comment]
> but if enough people don't like it they'll jump ship as soon as something better comes along that offers more freedom.

That would be great! If someone makes a tool that fits peoples’ workflows better than Prettier, we’d be happy to see it grow. See https://github.com/prettier/prettier/issues/3888#issuecomment-445063240 for a great explanation of why we don’t want to add options.

**Comment by slikts** [comment]
The issue is that Prettier tries to guess the more readable formatting based on a dumb heuristic: if it's even one character past a threshold, then wrap, otherwise, inline. This leads to Prettier sometimes guessing wrong, and also to diff churn. I think the best solution would be for Prettier to never inline, so that there's no diff churn, and for code editors and other tools displaying code to inline the code only for reading, and wrap the inline parts if they're selected for editing or feature a diff.

A more grounded solution is just to trade some consistency for more readability and less diff churns, or at least to make the heuristics more flexible.

**Comment by waynebloss** [comment]
A glaring inconsistency is being pointed out here though. So your explanation is actually proving the point of the issue being raised. In light of that explanation, a change is clearly needed to address this issue.

Also, the whole explanation rests on 2 completely subjective notions:

1. What is "readable".
2. That spending time think about how to format your code somehow "doesn't matter".

I'd call that a terrible explanation.

**Comment by j-f1** [comment]
> I think the best solution would be for Prettier to never inline

That’s already possible — just set `printWidth` to 1.

**Comment by slikts** [comment]
> That’s already possible — just set printWidth to 1.

Sure, but the point is that I'd still like my editor and code review tool to inline it for reading, just not when it's being edited or has changes.

**Comment by j-f1** [comment]
That sounds like a separate project, probably in the form of an editor plugin.

**Comment by soniabrami** [comment]
@j-f1 Is there any hope to have prettier keep multi-line formatting for destructuring assignment in the near future?

**Comment by j-f1** [comment]
I don’t think so.

**Comment by waynebloss** [comment]
I'm fine without it - not because I think that the lack of any options surrounding multi-line formatting is the right decision, but because it's entertaining to watch my code switch from multi-line to single line and back whenever I change a couple of characters and hit save... Spending the requisite time to find my place again always reminds me that "I'm doing this for the team" and "Maybe someday something better will come along or things will change".  Things are always changing, you know? Keeps things interesting.

Anyway, Prettier is certainly opinionated, but opinions can be changed. Perhaps if you want a better response, someone could make a funny YouTube video with Benny Hill theme music playing as you do a few find and replaces that cause the formatter to go crazy every time you save. Then post it to places like Hacker News, Reddit, Slashdot and elsewhere. It's not for me, but it could be fun.

**Comment by rcherny** [comment]
This discussion is so lengthy that it's hard to follow the current thinking here. I'm personally just wondering why destructured assignments don't follow the same rules as multi-line objects, as described in the docs. That doesn't imply a need for any new options.

The disconnect is odd to me, but I'm sure there's a reason the team doesn't want to do it. I see it doing it for "nested" objects, which is nice for sure.

[Single Line to Multi-line Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEB6AVAHSpmBZAVwBsYBLAWiNKjgAIAROAZxgCcCwYDXqBzbWoNoBJAHIBhAPKiAysJkAVAKKiFtAGasIAW1qESFKjVqSARgCs4nJrVNwiEAO7Z0qbKlQiokKE1IsEGFoAOlDaAAooCCCoUjA4AEpsHxZaYFpsXBgACz5aAF5aAHIigBpMmGjsuFYC2jYCOGwAXzrHCFYAEwBuTO9oVPSBIQ9aPyheIjojOk7mMHZObjhO0toaAM6I2PikqEFcKABDbTpC9Og4JHr2ODWYduv1I6ImOmby-ZHPbWIySmos2YDSWrBWaw28C2kTiiWGuAAbi9GnVgMMhLRLtcGnd0UIHhAni83niPhUcnlCiVPrgqjU6jiWm0Oj0+hgMjh8H9DICTBYrDAmBVsjAYAAHJhIDxisGi0g1YKkCCoToQMBMVAIVCsI5kaAvODBEXaIgAYl+BgBNHIEH51hcbigo14EGY9QgYz4U1oMzWpiOby20Fo7CmNjmTAWpDsWyOpggCLoMN2yQGQUcuXgqNoYu4YqmRNeuK+piIRzAAGtCySoM02Z4XW7KrQLf9fbYAytMftQ26I1GY7Q4wmkzs4f1fEFS+WK6jhqXGtjbp9BBnSPBq01a9gQKUQBAxXrfMhQEdWFpHAAFM8IJjIEBHBEQUidXcgUw6ytwGAyMXlvjIDie7GkQADqmbMH+8QyLe66kAi64AJ73uqd57tQbysDAl46rw2hHMgzxFnu5hMAAHgAQp+FbfjIJxwAAMoChHEncICkWRMhenAACKBDRHALHESAf6sJh97+nYRBvjK1AwKBL45MgAAcACse4yhAbygTqYr3jKzA1Imb4AI78fAOEHneKABuQNArCsb5gmZpBgjhRx4QRSBEW8e5vNopCAcuIDjJMvHmYJ3msXuMBxgpnRKUgABMMU6qQRi8OIOj4feUQ0G+BBvAocbWT5cDNM0QA)

**Comment by lydell** [comment]
> I'm personally just wondering why destructured assignments don't follow the same rules as multi-line objects, as described in the docs. 

As mentioned in the docs, object literals are used for a lot of different things in JavaScript and we haven't been able to find a good rule for all those cases, so we went with the "solution" of letting the user decide. But kind of the whole point of Prettier is _not_ letting the user decide on formatting things. Desctructuring, on the other hand, is used for destructuring only and using Prettier's normal linebreak rules works better (though not perfectly) there. Quoting the second comment of this thread:

> the special-casing of objects is more of a failing in prettier than a feature, I don't think it makes sense to spread that to more cases.

**Comment by duailibe** [comment]
> The issue is that Prettier tries to guess the more readable formatting based on a dumb heuristic: if it's even one character past a threshold, then wrap, otherwise, inline.

@slikts is correct, that is a dumb heuristic. However, it's the best heuristic we could find and make it work. And well, turns out it works fine for the majority of cases.

We'd love to see more tools that use other heuristics.

**Comment by slikts** [comment]
Any solution that conditionally inlines would cause diff churn, and inline editing will always be less convenient to edit; my ideal solution would be to only have *virtual* inlining in the editor while the actual formatting that's saved on disk would have no inlining, so that there's no diff churn. The virtual inlining would be based on line length like now, and also on cursor position, so moving the cursor to a line would  expand it to multiple lines.

**Comment by duailibe** [comment]
@slikts I'm looking forward to see other solutions for this problem.

**Comment by rclai** [comment]
Any progress as of late? I think the diff-friendliness is a strong argument for this feature.

**Comment by gkatsanos** [comment]
I started adding prettier to our workflow and I think this is important enough to rollback the adoption. A simple boolean option would suffice to turn off this heuristic.

**Comment by gkatsanos** [comment]
> > I think the best solution would be for Prettier to never inline
> 
> That’s already possible — just set `printWidth` to 1.

That forces ALL assignments to break into a new line. What we need is `printWidth: false`

**Comment by j-f1** [comment]
That’s not possible @gkatsanos — Prettier’s core algorithm depends on a defined `printWidth`.

**Comment by waynebloss** [comment]
Here is probably the least effort way of forcing prettier to do what you want:

```js
const {
  //
  x,
  y,
  z,
} = yadaYadaYada();
```

I find the maintainers' insistence that this option is not needed to be unusual, given that the same logic is not applied to the following functions at all.

If I wrote the function like this, prettier leaves it alone:

```js
function ({
  x,
  y,
  z
}) {
  return [x, y, z];
}
```

However, if I wrote the function like this, prettier also leaves it alone:

```js
function ({ x, y, z }) {
  return [x, y, z];
}
```

Can't we just have the same logic for destructuring already?

**Comment by thorn0** [comment]
> If I wrote the function like this, prettier leaves it alone:

Is it a proposal or a description of current behavior? If the latter, I can't reproduce it. Could you add links to the playground?

**Comment by waynebloss** [comment]
Haha - you know what? It was leaving it alone because it couldn't parse it. My mistake - I wasn't even paying attention to what I was typing.

(If you can't see the history of my comment - I basically forgot to type function names for my functions lol.)

**Comment by waynebloss** [comment]
Zero movement on this?

I have tons of code like this now and I'd really like to clean it up without having to fork this entire project:

```js
function myFunction({
  //
  xxx,
  yyy,
  zzz,
}) {
 ...
}
```

**Comment by Fi1osof** [comment]
JSON.stringify(object, undefuned, 2) do as i want. Just do like this, please. One-line - hell.

**Comment by balibebas** [comment]
It's hard to argue spreading a multi-liner onto a single line doesn't hurt one's ability to diff code. However, it also leaves code smell in some situations helping identify opportunities for refactoring like:

```jsx
{ container, header, body } = classes
```

Where you've got N identifiers in two places—not ideal even with multi-line as it'll impede refactoring. Switching to the following is therefore easier to maintain and Prettier will reward under the current setup:

```jsx
<Html className={classes.container}>
  <Head/>
  <Body className={classes.body}>
    <Heading className={classes.header}>
```

So there's that.

**Comment by mlshv** [comment]
Hello people! Is there any work happening on this feature? Is prettier going to make my code pretty? 🙂

**Comment by G1itcher** [comment]
I hate to do this but I'd really like to see an update on this one.

**Comment by ldaza-w** [comment]
![image](https://user-images.githubusercontent.com/77399493/153058632-b53bf55a-9fc1-48a8-bf78-709a25364335.png)

look at this... this just looks bad.. #uglier

**Comment by danramac** [comment]
This is not prettier

```
    const { text, hitSlop, onPress, style, activeOpacity, testID } =
      this.props;
```

**Comment by pvinis** [comment]
> Do you really

that's why prettier is a bad tool. because some people prefer the first, and some the second.

**Comment by jimfilippou** [comment]
> Do you really think this is prettier
> 
> ```
> const { text, hitSlop, onPress, style, activeOpacity, testID } =
>   this.props;
> ```
> 
> than this..
> 
> ```
> const {
>   text,
>   hitSlop,
>   onPress,
>   style,
>   activeOpacity,
>   testID 
> } = this.props;
> ```
> 
> ??

I do, I also respect @pvinis's opinion since the concept of "pretty" differs from person to person. 
Using an opinionated tool against your own opinion can be a determining factor in choosing prettier or not.

**Comment by amcinerney** [comment]
Keeping this alive. Yet another person who would like to see an option for this. I much prefer deconstruction/object fields to be on new lines for readability

**Comment by douglasg14b** [comment]
I to want an option for this, being able to say `Don't condense multi-line objects to a single line` would be nice to have, because there are some conventions where we WANT multi-line object,s but prettier insists that we cannot, with no clear way to disable or configure that behavior.

**Comment by gaievskyi** [comment]
> @Brian-Gaffney makes an excelent point, I use that destructuring a lot on react pure components and I'm a little tired of using `// prettier-ignore` plus I found myself on this scenario:
> 
> when you add or remove a key the diff will look like
> 
> ```diff
> import {
>   Foo,
>   Bar,
> -  Baz, 
> }
> ```
> 
> but with prettier the hole line has changes
> 
> ```diff
> - import { Foo, Bar, Baz } from './Qux'
> + import { Foo, Bar } from './Qux'
> ```
> 
> PS - Going for the minimal example here but keep in consideration real use cases like
> 
> ```
>   name: {
>     first,
>     last,
>   },
>   organisation: {
>     address: {
>       street: orgStreetAddress,
>       postcode: orgPostcode,
>     },
>   },
> }
> ```

That's actually _perfect_. I also still want to see an option for this.

**Comment by ZaLiTHkA** [comment]
at the risk of throwing a spanner in the works, I just had completely random thought: could Prettier perhaps format these sections conditionally based on the existence of explicit trailing commas?

---

drawing a little inspiration from the Dart formatter here: if code is written _with_ a final trailing comma, then the formatter puts each entry on it's own line. whereas _without_ this final trailing comma, it puts each entry on the same line (where applicable).

naturally, to **not** change current default behaviour (i.e.: usage with no explicit configuration), a change like this would need to be hidden behind an appropriate configuration flag though..

but I imagine a boolean-style "respect-trailing-comma" toggle could work quite nicely, where setting this to `true` would then format based on the existence of trailing commas that (_this tool should assume_) have been explicitly inserted by the dev.

---

FWIW, the following Dart snippet was all formatted in a single file:

```dart
var listExampleA = [
  'word',
  'more words',
  'final word',
];

var listExampleB = ['word', 'more words', 'final word'];

void funcExampleA(var arg1, var arg2) {}

void funcExampleB(
  var arg1,
  var arg2,
) {}
```

**Comment by adrian-gierakowski** [comment]
For anyone looking for an alternative formatter, check out https://github.com/dprint/dprint-plugin-typescript

it’s blazing fast and is reasonably configurable so that the choice of multi-line vs single line can be left to the developer

**Comment by waynebloss** [comment]
Here's another alternative that implements this feature: `Deno fmt`, the [Deno code formatter](https://deno.land/manual@v1.24.3/tools/formatter)

This can be used in VS Code via the [Deno extension](https://marketplace.visualstudio.com/items?itemName=denoland.vscode-deno). Just set your default formatter to it and you can leave the `deno.enable` VS Code setting `false` for your non-Deno projects.

Of course, if you are doing formatting in your CI/CD pipeline you just have to [install deno](https://deno.land/manual@v1.24.3/getting_started/installation) there too.

```json
{
"[typescript]": {
    "editor.defaultFormatter": "denoland.vscode-deno"
  },
}
```

**Comment by ZaLiTHkA** [comment]
I discovered a Prettier plugin earlier today through [this comment](https://github.com/prettier/prettier/issues/2716#issuecomment-1033923319), which covers my needs with regards to JS _arrays_, but not for _objects_, as required for this issue..

since I've never tried my hand at creating Prettier plugins before, I might be missing something.. but perhaps [`prettier-plugin-multiline-arrays`](https://github.com/electrovir/prettier-plugin-multiline-arrays) could be extended or forked to cover _this_ scenario as well?

**Comment by igorrocha** [comment]
I saw some people had issues in the past with multi-line arrays being put on a singleline by Prettier, and their workaround was to add a comment after the array's first element. I've tested with destructuring and it works too, even though it's not a great solution.

Writing your destructuring like this will keep it untouched by Prettier
```
const {
    prop1, //
    prop2,
    prop3,
} = props

**Comment by OJFord** [comment]
I do a similar thing for imports:
```
import {
  foo,
  //
} from "bar"
```

but it's annoying to need to.

It's not that this is necessarily 'prettier' than `{ foo }` itself, it's that the resulting _diffs_ are a hell of a lot prettier/more readable/more usable.

> The special-casing of objects is more of a failing in prettier than a feature

The failing is that it doesn't insist on readable diffs, and allows you to single-line it all if you remove the newline.

**Comment by Frankcarvajal** [comment]
Is it just me, or is it simply ridiculous that after 6 years this issue is still open and still there isn't multiline option for object deconstruction in prettier formatting with specificity for how many values are deconstructed before it breaks into multi line?

**Comment by adrian-gierakowski** [comment]
@Frankcarvajal 

prettier is dead

use dprint: https://github.com/prettier/prettier/issues/2550#issuecomment-1207867051

**Comment by bodenloszt** [comment]
> 

honestly :)). i was reading this entire issue and hoped to find the answer i need but well, surprise

**Comment by cpatti97100** [comment]
HI all, come on I think we can do this 🙏🏼 thanks again for this great tool :)

**Comment by sshmaxime** [comment]
Hi all, well, my turn now to wake things up 🙏

**Comment by superchangme** [comment]
> Hi all, well, my turn now to wake things up 🙏

has any new changes about this?
i use eslint rule
```
 // 建议6 对象字面量属性超过2个, 需要都换行
    'object-curly-newline': [
      'error',
      {
        ObjectExpression: {
          minProperties: 2,
          multiline: true,
          consistent: true
        },
        ObjectPattern: {
          minProperties: 6,
          multiline: true,
          consistent: true
        }
      }
```
but prettiter format me to one line，my brain want boom

**Comment by resistdesign** [comment]
I have this, and I hate it! I have to set printWidth down to like 50 to prevent it. 😨 
![image](https://github.com/prettier/prettier/assets/356375/ac5038e1-e06c-4467-a023-18d66a01d01f)

**Comment by CarlSalaets** [comment]
> at the risk of throwing a spanner in the works, I just had completely random thought: could Prettier perhaps format these sections conditionally based on the existence of explicit trailing commas?
> 
> drawing a little inspiration from the Dart formatter here: if code is written _with_ a final trailing comma, then the formatter puts each entry on it's own line. whereas _without_ this final trailing comma, it puts each entry on the same line (where applicable).
> 
> naturally, to **not** change current default behaviour (i.e.: usage with no explicit configuration), a change like this would need to be hidden behind an appropriate configuration flag though..
> 
> but I imagine a boolean-style "respect-trailing-comma" toggle could work quite nicely, where setting this to `true` would then format based on the existence of trailing commas that (_this tool should assume_) have been explicitly inserted by the dev.
> 
> FWIW, the following Dart snippet was all formatted in a single file:
> 
> ```dart
> var listExampleA = [
>   'word',
>   'more words',
>   'final word',
> ];
> 
> var listExampleB = ['word', 'more words', 'final word'];
> 
> void funcExampleA(var arg1, var arg2) {}
> 
> void funcExampleB(
>   var arg1,
>   var arg2,
> ) {}
> ```

This is a very underrated comment.

This gives predictable behaviour, and the flexibility to decide case by case.
It is a lot cleaner than the "comment hack" and it ties in perfectly with what a lot of people are doing anyway (i.e., using trailing comma's in multiline arrays/objects).

**Comment by ptrxyz** [comment]
This issue is open since 2017 and we keep on discussing it. I would highly advocate for implementing the trailing-comma-solution mentioned above  -- or any other one, really -- so that we move on for now.
Imho this doesn't need to be perfect, but at least it's better than nothing and we can improve it later on.

**Comment by DestinyZxw** [comment]
`	successHandle({
		item,
		storeItem,
		res,
		resolve
	}) {
		const {
			statusCode,
			data
		} = res;
		let bodyData;

		const {
			name,
			serverUrl
		} = storeItem

		try {
			bodyData = JSON.parse(data);
		} catch (e) {
			this.errHandle({
				item,
				storeItem,
				resolve,
			});
			return;
		}`

It's really cool！

**Comment by DestinyZxw** [comment]
successHandle({
		item,
		storeItem,
		res,
		resolve
	}) {
		const {
		    statusCode,
		    data
		} = res;
		let bodyData;

		const {
		    name,
		    serverUrl
		} = storeItem

		try {
			bodyData = JSON.parse(data);
		} catch (e) {
			this.errHandle({
				item,
				storeItem,
				resolve,
			});
		}
		// xxx more
	}	

really cool！

**Comment by resistdesign** [comment]
Huh? Did you use prettier on this somehow?

**Comment by Diomeh** [comment]
So it's been 7 years since this issue was opened with no solution on sight. I think it's pretty clear the developers wont ever fix this, even after solutions have been proposed.

Whoever reads this comment is, like me, forced to either put up with this issue or look elsewhere for a proper formatter.

**Comment by realing29** [comment]
> @Frankcarvajal
> 
> prettier is dead
> 
> use dprint: [#2550 (comment)](https://github.com/prettier/prettier/issues/2550#issuecomment-1207867051)

Could you tell me if there's destructuring discussed there, and if there's an analogue to the trivago prettier plugin's sort imports?

**Comment by adroslice** [comment]
We have objectWrap: preserve now, which I would expect to preserve these as well, but that doesn't seem to be the case. The only way I've seen to stop prettier from collapsing your object destructurings is to add line comments to at least one of your variables.
This issue gets a lot worse when fallback values and types are involved.
Real example from some svelte code:
```js
let {
  search = ((_) => []) as SearchFunc<T>,
  value = $bindable() as T,
} = $props();
```
Gets collapsed into:
```js
let { search = ((_) => []) as SearchFunc<T>, value = $bindable() as T } =
  $props();
```

I've considered opening a seperate issue, but I feel like that'll be closed as a dupe of this immediately despite the new context of the objectWrap option.

I must reiterate, **this is unexpected behavior and needs to be fixed**. Either preserve with an existing leading newline, or preserve with an existing trailing comma. Given that the latter is likely to cause problems when trailing commas are off, I would tend towards the former, and that also matches objectWraps behavior for objects.

**Comment by baweaver** [comment]
One more on the list of +1s here. Prettier is overly aggressive in collapsing lines, and we have near decade old issue highlighting this problem. Is there a way we can place a bounty on this feature to get interest on it? What amount would be sufficient to get this prioritized?

EDIT: This appears to be a vehicle to facilitate donations:

https://thanks.dev/u/gh/prettier

So to the maintainers: What would a reasonable amount be to look into this issue? I'm willing to start it out by offering $100 in matching to get things going for anyone else who wants to chip in.

**Comment by baweaver** [comment]
In the spirit of "being the change" I gave it a swing:

https://github.com/prettier/prettier/pull/18155

JS isn't my best language, so feedback more than welcome.

**Comment by Dunqing** [comment]
**Prettier 3.7.4**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEA3AhgJwEtcAjAGzkwF5NgAdKTTAQQYF8GHIoNbMGW7ajlyF8pCiAA0ICAAcYhdMlAF8EAO4AFAgjTIQuMhtwBPfTJLiwAazgwAyrgC2cADKEocZADMjaOBkIEgArODAYAHVxOQM5fDgA-GxvS2s7RzlcME8Ac2QYfABXQJAA50IC4tK4AA85OCJXWCMAeQbxGAh8LQg0QkVoAwQAE2kQOo7CZpgjABVGqAJCRN9-Uv6oXIoARSKIeDWyAJkQtFqHPN39w6Q-Y9KARxu4LXU5fRRcNABaLzgRgDxoVRGQ8gBhCDOZy4AxGMjjTbbODMGCFQgkIrwLSNDxeI4nEAACxgzjIkSJA0SWTAcAcegGhGwA1MBjAaAsIGwJQAklBAbAHGAiApmPyHDBTJI7usZPE+nBorhYih4olGilxqNWj48d4UGQfONPEkYG9cLkYQTSll8EkDCRSHAEXKiLBIoQRjAicgABwABhkCWeYle4ktsJlDxksxIHq9PqQACYZEUAnNSJ97oS4M4SADASM3LgtkULXAAGLdGFovJwrEQEBsNhAA)

```sh
--parser babel
--object-wrap preserve
```

**Input:**

```jsx
const variable = {
  A
}

const { 
 A
} = vairable
```

**Output:**

```jsx
const variable = {
  A,
};

const { A } = vairable;

```

**Expected output:**

```jsx
const variable = {
  A,
};

// Same as above, as there is a line break after `{`
const { 
  A
} = vairable;
```

**Why?**

I think the `objectWrap` behavior should also apply to the object pattern. That makes more sense for the sake of consistency, as they are all object-like structures. Ideally, there should be no breaking changes for any existing code, but this issue should be resolved.

**Comment by DaltonPelkey** [comment]
+1 - object destructuring in JSX/TSX is too common. Often times I am destructuring 5+ values from the return of a hook or component props. It is far more readable to have it break it into multiple lines. Especially if I explicitly write it with newlines starting from the opening bracket.
